runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -5278,3 +5282,103 @@ J9::CodeGenerator::stressJitDispatchJ9MethodJ2I()
    static const bool stress = feGetEnv("TR_stressJitDispatchJ9MethodJ2I") != NULL;
    return stress;
    }
+
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+void
+J9::CodeGenerator::addInvokeBasicCallSiteImpl(
+   TR::Node *callNode, TR::Instruction *instr, uint8_t *retAddr)
+   {
+   TR_ASSERT_FATAL_WITH_NODE(
+      callNode,
+      (instr != NULL) != (retAddr != NULL),
+      "expected exactly one of TR::Instruction or return address");
+
+   if (comp()->getOption(TR_TraceCG))
+      {
+      traceMsg(comp(), "Call instruction ");
+      if (instr != NULL)
+         {
+         traceMsg(comp(), "%p", instr);
+         }
+      else
+         {
+         traceMsg(comp(), "with return address %p", retAddr);
+         }
+
+      traceMsg(
+         comp(),
+         " for n%un [%p] may target VM MethodHandle.invokeBasic\n",
+         callNode->getGlobalIndex(),
+         callNode);
+      }
+
+   TR_J9VMBase *fej9 = comp()->fej9();
+   TR::Node *j2iCallNode = callNode;
+   TR::MethodSymbol *methodSymbol = callNode->getSymbol()->castToMethodSymbol();
+   TR::RecognizedMethod rm = methodSymbol->getMandatoryRecognizedMethod();
+   switch (rm)
+      {
+      case TR::java_lang_invoke_MethodHandle_invokeBasic:
+         break; // ok
+
+      case TR::com_ibm_jit_JITHelpers_dispatchVirtual:
+         j2iCallNode =
+            fej9->getEquivalentVirtualCallNodeForDispatchVirtual(callNode, comp());
+         break;
+
+      default:
+         TR_ASSERT_FATAL_WITH_NODE(
+            callNode,
+            false,
+            "expected MethodHandle.invokeBasic or JITHelpers.dispatchVirtual");
+         break;
+      }
+
+   int32_t numChildren = j2iCallNode->getNumChildren();
+   int32_t firstArgIndex = j2iCallNode->getFirstArgumentIndex();
+   uint32_t numArgSlots32 = 0;
+   for (int32_t i = firstArgIndex; i < numChildren; i++)
+      {
+      TR::Node *child = j2iCallNode->getChild(i);
+      TR::DataTypes dt = child->getDataType().getDataType();
+      numArgSlots32 += 1 + (uint32_t)(dt == TR::Int64 || dt == TR::Double);
```

Comment:
Sorry if this is a dumb question, but does this have to consider the possibility that the child node is a `TR::PassThrough`?

Reply:
Thanks, it will have to check for that, unfortunately